cmd/convertor/builder/builder.go

```diff
@@ -95,41 +97,167 @@ func NewOverlayBDBuilder(ctx context.Context, opt BuilderOptions) (Builder, erro
-	engineBase.dumpManifest = opt.DumpManifest
-
-	var engine builderEngine
-	switch opt.Engine {
-	case Overlaybd:
-		engine = NewOverlayBDBuilderEngine(engineBase)
-	case TurboOCI:
-		engine = NewTurboOCIBuilderEngine(engineBase)
+	if len(manifestDescs) == 0 {
+		log.G(ctx).Warn("no matched manifest, do nothing")
+		return nil
 	}
-	return &overlaybdBuilder{
-		layers: len(engineBase.manifest.Layers),
-		engine: engine,
-		config: engineBase.config,
-	}, nil
+
+	var mu sync.Mutex
+	var targetDescs []v1.Descriptor
+	g, gctx := errgroup.WithContext(ctx)
+	for _, _mDesc := range manifestDescs {
+		mDesc := _mDesc
+		g.Go(func() error {
+			var platform string
+			var rctx context.Context
+			if mDesc.Platform == nil {
+				platform = ""
+				rctx = gctx
+			} else {
+				platform = platforms.Format(*mDesc.Platform)
+				rctx = log.WithLogger(gctx, log.G(ctx).WithField("platform", platform))
+			}
+			workdir := filepath.Join(opt.WorkDir, strings.ReplaceAll(platform, "/", "-"))
+
+			log.G(rctx).WithFields(log.Fields{
+				"manifest digest": mDesc.Digest,
+				"work dir":        workdir,
+			}).Info("building ...")
+
+			// init build engine
+			fetcher, err := resolver.Fetcher(rctx, opt.Ref)
+			if err != nil {
+				return fmt.Errorf("failed to new fetcher: %w", err)
+			}
+			targetRef := opt.TargetRef
+			if len(manifestDescs) > 1 {
+				targetRef += "@" // append '@' to avoid tag
+			}
+			pusher, err := resolver.Pusher(rctx, targetRef)
+			if err != nil {
+				return fmt.Errorf("failed to new pusher: %w", err)
+			}
+			manifest, config, err := fetchManifestAndConfig(ctx, fetcher, mDesc)
+			if err != nil {
+				return fmt.Errorf("failed to fetch manifest and config: %w", err)
+			}
+			engineBase := &builderEngineBase{
+				resolver:  resolver,
+				fetcher:   fetcher,
+				pusher:    pusher,
+				manifest:  *manifest,
+				config:    *config,
+				inputDesc: mDesc,
+			}
+			engineBase.workDir = workdir
+			engineBase.oci = opt.OCI
+			engineBase.mkfs = opt.Mkfs
+			engineBase.vsize = opt.Vsize
+			engineBase.db = opt.DB
+			refspec, err := reference.Parse(opt.Ref)
+			if err != nil {
+				return err
+			}
+			engineBase.host = refspec.Hostname()
+			engineBase.repository = strings.TrimPrefix(refspec.Locator, engineBase.host+"/")
+			engineBase.reserve = opt.Reserve
+			engineBase.noUpload = opt.NoUpload
+			engineBase.dumpManifest = opt.DumpManifest
+
+			var engine builderEngine
+			switch opt.Engine {
+			case Overlaybd:
+				engine = NewOverlayBDBuilderEngine(engineBase)
+			case TurboOCI:
+				engine = NewTurboOCIBuilderEngine(engineBase)
+			}
+
+			// build
+			builder := &overlaybdBuilder{
+				layers: len(engineBase.manifest.Layers),
+				engine: engine,
+			}
+			desc, err := builder.Build(rctx)
+			if err != nil {
+				return fmt.Errorf("(platform %s) failed to build: %w", platform, err)
+			}
+			desc.Platform = mDesc.Platform
+
+			mu.Lock()
+			defer mu.Unlock()
+			targetDescs = append(targetDescs, desc)
```

Comment:
Are there any annotations that we could add to denote that the artifacts are streamable ?

Reply:
I made a few changes to keep the artifacts in their original order